### PR TITLE
[`core`] PEFT refactor + introducing `create_and_replace` public method

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do ?

Refactors a bit the internals of some PEFT models (making some methods static, etc.) and introduces a new method `create_and_replace` for users that want to pass a bare model, a peft config to inject adapters in-place into the model. 
These changes are totally BC with the previous PEFT versions.

This PR makes things easier for the PEFT integration in transformers https://github.com/huggingface/transformers/pull/25077 

cc @pacman100 @BenjaminBossan

### CODE REVIEW

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for much for working on this. IMO this is moving PEFT in the right direction (not only saying this because it brings it closer to my refactoring proposal ;-), even regardless of the transformers integration.

### general comments

> I am happy to break down into multiple smaller PRs

Yeah, there is a bit of diff noise because somewhat unrelated things were changed, but splitting into separate PRs is extra work, so I'm fine with keeping it as is.

> To tackle the main motivation I propose to differentiate things between two type of adapters

Yes, big +1 from my side. The two types are quite different and lots of code is basically a big if-else that does completely different things based on the type.

### create_and_replace

I think the biggest change of the PR is the new function `peft.mapping.create_and_replace`. IIUC, the purpose of that function is to create an adapted module with corresponding technique (lora, ia³, etc.) without any wrapper model around it. I agree that this is a good feature.

However, I think this somewhat duplicates existing functionality. At least with lora, we can create a `LoraModel` and then `unload` it, which achieves the same goal. I believe it would be better if we create an instance of the `LoraModel` and then return the unloaded model. That way, there is no duplication with the different `_find_and_replace` methods. I'm still in favor of factoring out that code, because it's mostly duplicated between the different adapter methods, but maybe just as a helper function.

So in sum, I think we can keep the `create_and_replace` function, but in its body, it would instantiate a `LoraModel` / `IA3Model` etc. based on the passed config, and then return `model.unload()`, so it would be a very short function that leverages existing code. WDYT?

Of course, we would have to support unloading on the non-LoRA methods too, but they should have it anyway (maybe we can add this method, and all the other missing ones, to the abstract base class?).

### Backwards compatibility

If I'm not missing something, this is indeed fully backwards compatible. Just to be sure though, I would test that a model created with the new method can indeed load the weights from an existing model.

In the long run, we could even add this kind of test to the test suite (regression testing) but it's probably out of scope for this PR.

### Taking this further

Also probably out of scope, but I'm just trying to imagine how we can take this direction even further. If we agree on this direction, that may inspire this PR as well.

IMHO, we should consider if we want to offer 2 APIs. One is a "high level" API and the other a "low level" API. Both are public, so we expect that advanced users may use the low level API.

The distinction would be that the high level API implements creation of PEFT models based on a PEFT config. To do that, the high level API would call the low level API. This would include an API to set individual adapter layers with specific parameters derived from the config.

That way, users can use the high level API the same way they do right now and create a PEFT model from config. However, more advanced users can create more fine-grained adapters, such as:

- lora layers with different values for `r` (see #752)
- mixing different types, like lora + ia³
- providing custom adapter layers written by the user

Obviously, the low level API cannot be represented by a config, since it is too flexible. That means, a config, via high level API, can map to low level API, but not vice versa. I think this is okay, though, because configs are always very restrictive in their expressiveness, and we shouldn't prevent advanced users to do more powerful things.

Another advantage of this better separation of concerns would be that if someone wants to implement a new tuning method, they only have to implement the low level API. That means less code duplication, like we have right now with the finding and replacement layers that is duplicated for each method.

But yeah, as I mentioned, that's out of scope for this PR, just my flights of fancy :)

File: src/peft/peft_model.py [hunk 1]
```diff
@@ -531,56 +530,13 @@ def load_adapter(self, model_id: str, adapter_name: str, is_trainable: bool = Fa
...
-                        f"Please check that the file {WEIGHTS_NAME} or {SAFETENSORS_WEIGHTS_NAME} is present at {model_id}."
-                    )
-
-        if use_safetensors:
-            adapters_weights = safe_load_file(filename, device="cuda" if torch.cuda.is_available() else "cpu")
-        else:
-            adapters_weights = torch.load(
-                filename, map_location=torch.device("cuda" if torch.cuda.is_available() else "cpu")
-            )
+        adapters_weights = self._get_peft_state_dict(model_id, **hf_hub_download_kwargs)
```
Comment (pacman100) [inline src/peft/peft_model.py:539]: as `_get_peft_state_dict` is loading weights, the naming here is confusing. It should be `_load_peft_weights` or something along those lines

**Review by pacman100 — COMMENTED** [review]
Hello @younesbelkada, thank you for all the great efforts are putting into this! 🤗

My major concern is that `create_and_replace` is leading to lot of duplication of code. One way to overcome it is to use `create_and_replace` in `_find_and_replace`.

File: src/peft/peft_model.py (hunk 1 above)
Comment (younesbelkada) [reply src/peft/peft_model.py]: Agreed, I have moved the method to `save_and_load` utils and renamed it

Comment (younesbelkada) [reply src/peft/peft_model.py]: Also made the method public so that it can be used for the transformers integration and more generally by users

File: src/peft/mapping.py [hunk 2]
```diff
@@ -89,10 +90,45 @@ def get_peft_model(model: PreTrainedModel, peft_config: PeftConfig, adapter_name
...
+
+# TODO: docstring and typehints
+def create_and_replace(peft_type, peft_config, model, adapter_name):
+    if peft_type not in TUNERS_MAPPING:
+        raise ValueError(
+            f"Task type {peft_type} is not supported. Supported task types are {list(TUNERS_MAPPING.keys())}"
+        )
+    tuner_cls = TUNERS_MAPPING[peft_type]
+
+    for _, param in model.named_parameters():
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/peft/mapping.py:108]: I can confirm the suggestion above is sufficient

File: src/peft/tuners/adalora.py [hunk 3]
```diff
@@ -121,6 +121,102 @@ def add_adapter(self, adapter_name, config=None):
...
+            "lora_alpha": lora_config.lora_alpha,
+            "lora_dropout": lora_config.lora_dropout,
+            "fan_in_fan_out": lora_config.fan_in_fan_out,
+            "init_lora_weights": lora_config.init_lora_weights,
+            "loaded_in_8bit": loaded_in_8bit,
+            "loaded_in_4bit": loaded_in_4bit,
+        }
+
+        # If it is not a LoraLayer, create a new module, else update it with new adapters
+        if not isinstance(target, LoraLayer):
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/peft/tuners/adalora.py]: Yes, changed it !

File: src/peft/tuners/tuners_utils.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+            The name of the active adapter.
+    """
+    _is_plugable = False
+    active_adapter = None
+
+    @property
+    def peft_is_plugable(self):
+        return self._is_plugable
+
+    def __post_init__(self):
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/peft/tuners/tuners_utils.py:124]: I used `_ peft_is_plugable` in the transformers integration but I will remove it, I will propose something else for the post init method

File: src/peft/tuners/adalora.py [hunk 5]
```diff
@@ -121,6 +121,102 @@ def add_adapter(self, adapter_name, config=None):
             self.trainable_adapter_name = adapter_name
             self.rankallocator = RankAllocator(self.model, self.peft_config[adapter_name], self.trainable_adapter_name)
 
+    @staticmethod
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/peft/tuners/adalora.py:139]: Sounds great ! Thanks for this suggestion

**Review by BenjaminBossan — COMMENTED** [review]
I was going to do an in-depth review but I struggled a bit with understanding how the different pieces fit together. Therefore, I created a schema, which is shown in the figure below. Please let me know if the schema is incorrect.

While looking at it, I couldn't stop but feel like the way things are connected is more complex than necessary. Therefore, I drafted an alternative suggestion at the bottom of the figure (for the case of LoRA but the same should work with others). Please let me know if it makes sense.

Basically, I think the alternative is simpler while achieving the same goal. A main difference is that with the current suggestion, when calling from transformers, there is never a `PeftModel` or `LoraModel` instance being created (because only class and static methods are called). However, I'm not sure why that would be an issue. In my suggestion, we create an instance but in the end only return the `base_model`. Is there some important reason I'm missing why that cannot work?

schema-refactor

**Review by younesbelkada — COMMENTED** [review]
Thanks @BenjaminBossan for the valuable feedback and clear picture. I should have adapted the design accordingly with a minor tweak, I propose to define an instance method `_create_and_replace` for each subclass since I prefer to keep the `_replace_module` method as it is so that it can be used correctly by AdaLoraModel without code de-duplication. Also that method was used by the `_unload_and_optionally_merge` method.
I have also fixed AdaLora failing tests. Would like to hear from you about this design!

File: src/peft/config.py [hunk 6]
```diff
@@ -102,6 +82,9 @@ def from_pretrained(cls, pretrained_model_name_or_path, subfolder=None, **kwargs
             kwargs (additional keyword arguments, *optional*):
                 Additional keyword arguments passed along to the child class initialization.
         """
+        # Avoid circular dependency .. TODO: fix this with a larger refactor
```
Comment (BenjaminBossan) [inline src/peft/config.py:85]: Most of the time, circular imports means that the Python modules or not well organized. I haven't checked exactly what the issue is here but maybe it would be addressed by the comment you made earlier today in the meeting.

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
_Regarding the failing tests, I haven't checked yet what the issue is, will likely do it later, but I wanted to give feedback as early as possible, so here it is :)_

I like these recent changes. I think this is a good clean up of the code with some refactoring and better definition of interfaces. Overall, I also have more confidence now that this code is doing the correct thing.

Since quite a few changes were about moving code from A to B, I have a small request: Let the reviewers know which code was moved unchanged, and which code was moved + changed, so that we know what we need to check and what we can skip.

I don't have too many comments left. A few I made, I believe, should help to simplify the code even further. Please take a look if they makes sense to you.

File: src/peft/mapping.py [hunk 7]
```diff
@@ -89,10 +90,35 @@ def get_peft_model(model: PreTrainedModel, peft_config: PeftConfig, adapter_name
...
+
+    Args:
+        (`peft_config`, `PeftConfig`):
+            Configuration object containing the parameters of the Peft model.
+        (`model`, `torch.nn.Module`):
+            The input model where the adapter will be injected.
+        (`adapter_name`, `str`):
+            The name of the adapter to be injected.
+    """
+    if not isinstance(peft_config, PeftConfig):
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/peft/mapping.py:115]: Not checking the type sounds good also !

File: src/peft/peft_model.py [hunk 8]
```diff
@@ -531,56 +528,13 @@ def load_adapter(self, model_id: str, adapter_name: str, is_trainable: bool = Fa
...
-                        f"Please check that the file {WEIGHTS_NAME} or {SAFETENSORS_WEIGHTS_NAME} is present at {model_id}."
-                    )
-
-        if use_safetensors:
-            adapters_weights = safe_load_file(filename, device="cuda" if torch.cuda.is_available() else "cpu")
-        else:
-            adapters_weights = torch.load(
-                filename, map_location=torch.device("cuda" if torch.cuda.is_available() else "cpu")
-            )
+        adapters_weights = load_peft_weights(model_id, **hf_hub_download_kwargs)
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/peft/peft_model.py:540]: Thanks !

File: src/peft/tuners/tuners_utils.py [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+            self.peft_config = peft_config
+
+        # transformers models have a .config attribute, whose presence is assumed later on
+        if not hasattr(self, "config"):
+            self.config = {"model_type": "custom"}
+
+        self.adapter_layer_class = adapter_layer_class
+        self.create_and_replace(self.model, adapter_name)
+
+    def pre_init(self, model, config, adapter_name, adapter_layer_class):
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/peft/tuners/tuners_utils.py:47]: sounds great let's remove these checks

File: src/peft/utils/save_and_load.py [hunk 10]
```diff
@@ -118,14 +126,71 @@ def set_peft_model_state_dict(model, peft_model_state_dict, adapter_name="defaul
...
     load_result = model.load_state_dict(peft_model_state_dict, strict=False)
-    if isinstance(config, PromptLearningConfig):
+    if config.is_prompt_learning:
         model.prompt_encoder[adapter_name].embedding.load_state_dict(
             {"weight": peft_model_state_dict["prompt_embeddings"]}, strict=True
         )
     return load_result
+
+
+def load_peft_weights(model_id, **hf_hub_download_kwargs):
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/peft/utils/save_and_load.py:142]: Yes that's correct

File: src/peft/tuners/ia3.py [hunk 11]
```diff
@@ -91,7 +92,40 @@ def __post_init__(self):
         self.peft_type = PeftType.IA3
 
 
-class IA3Model(torch.nn.Module):
+class IA3Layer(BaseTunerLayerMixin):
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/peft/tuners/ia3.py:95]: Correct yes

File: src/peft/tuners/adalora.py [hunk 12]
```diff
@@ -61,6 +60,50 @@ def __post_init__(self):
         self.peft_type = PeftType.ADALORA
 
 
+class AdaLoraLayer(LoraLayer):
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/peft/tuners/adalora.py:63]: Correct yes!

File: src/peft/tuners/adalora.py [hunk 13]
```diff
@@ -88,17 +131,8 @@ class AdaLoraModel(LoraModel):
...
-        self.peft_config = config
-        self.add_adapter(adapter_name, self.peft_config[adapter_name])
-
-    def add_adapter(self, adapter_name, config=None):
-        if config is not None:
-            model_config = self.model.config.to_dict() if hasattr(self.model.config, "to_dict") else self.model.config
-            config = self._prepare_adalora_config(config, model_config)
-            self.peft_config[adapter_name] = config
-        self._find_and_replace(adapter_name)
+        BaseTuner.__init__(self, model, config, adapter_name, adapter_layer_class=AdaLoraLayer)
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/peft/tuners/adalora.py]: Agreed with all the suggestions here, makes sense!

File: src/peft/tuners/tuners_utils.py [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+    A tuner layer mixin that provides the common methods and attributes for all tuners.
+
+    Args:
+        is_plugable (`bool`, *optional*):
+            Whether the adapter layer can be plugged to any pytorch module
+        active_adapter (`str`, *optional*):
+            The name of the active adapter.
+    """
+    active_adapter = None
+    supports_merging = False
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/peft/tuners/tuners_utils.py]: removed

File: src/peft/tuners/adalora.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/peft/tuners/adalora.py]: Should have addressed them in [a92e5ea](https://github.com/huggingface/peft/pull/749/commits/a92e5ea8863cb12f8d49438d124c161cc47ce87e)

File: src/peft/tuners/tuners_utils.py [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+        """
+        if not self.adapter_layer_class.supports_merging:
+            raise ValueError(f"{self.__class__.__name__} does not support merging adapter layers")
+
+        for module in self.model.modules():
+            if isinstance(module, self.adapter_layer_class):
+                module.unmerge()
+
+
+class BaseTunerLayerMixin:
```
[the remark this answers is not in the document]
Comment (BenjaminBossan) [reply src/peft/tuners/tuners_utils.py]: Hmm yes, you're right, I thought for some reason it had `merge` and `unmerge` too, but it doesn't. So disregard that comment. I do wonder if `ModulesToSaveWrapper` should have those methods, though.

File: src/peft/tuners/tuners_utils.py [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+        if adapter_layer_class is None:
+            raise ValueError("You need to specify a `adapter_layer_class` attribute in your class")
+        if not isinstance(adapter_name, str):
+            raise ValueError("You need to specify a `adapter_name` attribute in your class")
+        if not isinstance(model, nn.Module):
+            raise ValueError("The model attribute needs to be a `nn.Module` instance")
+
+    @abstractmethod
+    def _prepare_adapter_config(self, peft_config, model_config):
+        ...
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/peft/tuners/tuners_utils.py:61]: Added docstrings and nice comments in [45e8877](https://github.com/huggingface/peft/pull/749/commits/45e8877f7484ea292e13055e66f22753e2dd0736)

File: src/peft/config.py (hunk 6 above)
Comment (younesbelkada) [reply src/peft/config.py:85]: Yes I think moving the tuner classes into separate folders should help

File: src/peft/tuners/tuners_utils.py [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+
+    @abstractmethod
+    def _mark_only_adapters_as_trainable(self):
+        ...
+
+    def create_and_replace(self, model, adapter_name):
+        peft_config = self.peft_config[adapter_name]
+
+        if not isinstance(peft_config, PeftConfig):
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/peft/tuners/tuners_utils.py]: We don't need it indeed

File: src/peft/tuners/tuners_utils.py [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+            raise ValueError(f"peft_config must be an instance of PeftConfig got {type(peft_config)} instead.")
+
+        # TODO: test that
+        for module in model.modules():
+            if isinstance(module, BaseTunerLayerMixin):
+                module.requires_grad_(False)
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/peft/tuners/tuners_utils.py]: I can confirm this is not needed as `_mark_only_adapters_as_trainable` will also freeze the non-adapter weights later on

File: src/peft/config.py [hunk 19]
```diff
@@ -122,7 +105,13 @@ def from_pretrained(cls, pretrained_model_name_or_path, subfolder=None, **kwargs
 
         loaded_attributes = cls.from_json_file(config_file)
 
-        config = cls(**class_kwargs)
+        if "peft_type" in loaded_attributes:
+            peft_type = loaded_attributes["peft_type"]
+            config_cls = PEFT_TYPE_TO_CONFIG_MAPPING[peft_type]
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/peft/config.py:124]: Added some comments in [4ce8541](https://github.com/huggingface/peft/pull/749/commits/4ce8541c052c85d54bd0f9fa244bfad00fbd7f85) to explain the reason of these changes

File: src/peft/tuners/lora.py [hunk 20]
```diff
@@ -117,7 +118,100 @@ def __post_init__(self):
         self.peft_type = PeftType.LORA
 
 
-class LoraModel(torch.nn.Module):
+# This class has just been moved above
```
Comment (BenjaminBossan) [inline src/peft/tuners/lora.py:121]: I don't think this type of comment is necessary. I just meant it should be included in the description of the PR for the reviewers to make their lives easier :)

Also, the comment is not quite true because `LoraLayer` has been changed a little bit :D

File: src/peft/tuners/tuners_utils.py [hunk 21]
```diff
@@ -0,0 +1,203 @@
...
+
+    def __init__(self, model, peft_config, adapter_name):
+        super().__init__()
+
+        self.model = model
+        self.forward = self.model.forward
+
+        # Some adapters might be already attached
+        if not hasattr(self, "peft_config"):
+            self.peft_config = peft_config
```
Comment (BenjaminBossan) [inline src/peft/tuners/tuners_utils.py:37]: I am uncertain about when the `else` case could happen: If there is already a `peft_config` attribute, the `peft_config` being passed here is just ignored. Is this something that could be confusing to users? Or maybe we should check that the existing `peft_config` is identical to the one being passed here and issue a warning if not?

File: src/peft/mapping.py [hunk 22]
```diff
@@ -89,10 +90,30 @@ def get_peft_model(model: PreTrainedModel, peft_config: PeftConfig, adapter_name
...
-    if isinstance(peft_config, PromptLearningConfig):
+    if peft_config.is_prompt_learning:
         peft_config = _prepare_prompt_learning_config(peft_config, model_config)
     return MODEL_TYPE_TO_PEFT_MODEL_MAPPING[peft_config.task_type](model, peft_config, adapter_name=adapter_name)
+
+
+def create_and_replace(peft_config: PeftConfig, model: torch.nn.Module, adapter_name: str):
+    r"""
+    A simple API to create and inject adapter in-place into a model. Currently the API does not support prompt learning
+    methods and only LoRA, AdaLora and IA3 are supported. Make sure to have the correct `target_names` set in the
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/peft/mapping.py:103]: Makes sense!

File: src/peft/mapping.py [hunk 23]
```diff
@@ -89,10 +90,30 @@ def get_peft_model(model: PreTrainedModel, peft_config: PeftConfig, adapter_name
...
+    `peft_config` object. The API calls `get_peft_model` under the hood but would be restricted only to non-prompt
+    learning methods.
+
+    Args:
+        (`peft_config`, `PeftConfig`):
+            Configuration object containing the parameters of the Peft model.
+        (`model`, `torch.nn.Module`):
+            The input model where the adapter will be injected.
+        (`adapter_name`, `str`):
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/peft/mapping.py]: ah correct, you are right, just changed it in c3e6b61

File: src/peft/tuners/lora.py (hunk 20 above)
Comment (younesbelkada) [reply src/peft/tuners/lora.py:121]: makes sense :D I thought you meant to explicitly add comments about that, removed it!

File: src/peft/tuners/tuners_utils.py [hunk 24]
```diff
@@ -0,0 +1,203 @@
...
+from torch import nn
+
+from ..config import PeftConfig
+from ..utils import _get_submodules
+
+
+class BaseTuner(nn.Module, ABC):
+    r"""
+    A base tuner model that provides the common methods and attributes for all tuners. Each `BaseTuner` class needs to
+    be implemented in case the adapter is a plugable adapter.
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/peft/tuners/tuners_utils.py]: Just refactored the docstring of that class in [4d1dffb](https://github.com/huggingface/peft/pull/749/commits/4d1dffbb10b43f0b734bdb28c5198a76b948d48e)

File: src/peft/tuners/tuners_utils.py (hunk 21 above)
Comment (younesbelkada) [reply src/peft/tuners/tuners_utils.py:63]: Hmm actually this condition is not needed for PEFT but I added it as a hack to make it work for the transformers integration + multiple adapters. 

`model.load_adapter` will re-create new PeftModel instance and change the current model in-place. If I don't add that hack it will overwrite the current existing peft_config dict with a new one

Perhaps I can elaborate this in the comments or add a warning telling users that there is already a `peft_config` attached to the model ?

Comment (BenjaminBossan) [reply src/peft/tuners/tuners_utils.py:69]: I see. So if this is the only situation where this occurs, I think a comment is enough.

Comment (younesbelkada) [reply src/peft/tuners/tuners_utils.py:64]: Added an info and more comments

File: src/peft/tuners/adaption_prompt.py [hunk 25]
```diff
@@ -163,10 +168,10 @@ def __init__(self, model, configs: Dict, adapter_name: str):
         # Whether the adapter is enabled.
         self._enabled = True
         self.forward = self.model.forward
-        self.add_adapter(adapter_name, configs[adapter_name])
+        self._add_adapter(adapter_name, configs[adapter_name])
         self._mark_only_adaption_prompts_as_trainable()
 
-    def add_adapter(self, adapter_name: str, config: AdaptionPromptConfig) -> None:
+    def _add_adapter(self, adapter_name: str, config: AdaptionPromptConfig) -> None:
```
[the remark this answers is not in the document]
Comment (BenjaminBossan) [reply src/peft/tuners/adaption_prompt.py:174]: The recursion error occurs because in `PeftModel`, it should be `self.base_model.add_adapter(adapter_name, peft_config)`, otherwise the method calls itself. The reason why renaming works is because of the `__getattr__` of `PeftModel` delegates to `base_model` (this confusion is btw. a good case in point why `__getattr__` should usually not be overridden).

Apart from that, it should not be necessary to rename the method, I tried it out locally and it works with the old name.

**Review by BenjaminBossan — COMMENTED** [review]
From my point of view, almost perfect, only two small comments. Thanks @younesbelkada

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot @BenjaminBossan for the amazing review and for your patience - things should be much smoother now!

**Comment by younesbelkada** [comment]
Thanks very much @BenjaminBossan for your extensive review and all your patience ! 
@pacman100 I can also confirm all slow tests pass on this branch, validating the 4bit / 8bit compatibility of this PR

**Review by pacman100 — COMMENTED** [review]
Hello @younesbelkada, Thank you for iterating and adding `BaseTuner` and `BaseTunerLayer` classes to standardize the injectable peft methods 😄. 

Currently, I am very confused with 3 functions/methods doing different things with name `create_and_replace.` There is an external function in utils called `create_and_replace`, `BaseTuner` has public method called `create_and_replace` as well as a private method `_create_and_replace` while all doing very different things. This makes it difficult to understand and something that is complicated when it shouldn't be the case. 

How about the following:
1. I feel the util function of `mapping.py` named `create_and_replace` is not required at all. See the comment on transformers PR: https://github.com/huggingface/transformers/pull/25077#discussion_r1281484623
2. Rename the public method of `BaseTuner` named `create_and_replace` to `add_adapter` or `inject_adapter` or `find_create_and_replace`, WDYT?

File: src/peft/tuners/ia3.py [hunk 26]
```diff
@@ -23,17 +23,18 @@
...
 from ..import_utils import is_bnb_available
 from ..utils import (
     TRANSFORMERS_MODELS_TO_IA3_FEEDFORWARD_MODULES_MAPPING,
     TRANSFORMERS_MODELS_TO_IA3_TARGET_MODULES_MAPPING,
     ModulesToSaveWrapper,
-    PeftConfig,
     PeftType,
-    _freeze_adapter,
     _get_submodules,
+    _is_valid_match,
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/peft/tuners/ia3.py:34]: Ah good point, I thought other tuner methods could benefit from it (since the method looks quite generic) in the future so I moved it into the utils

File: src/peft/tuners/ia3.py [hunk 27]
```diff
@@ -213,75 +217,67 @@ def _create_new_module(self, ia3_config, adapter_name, target, is_feedforward):
...
-
-    def _replace_module(self, parent_module, child_name, new_module, old_module):
-        setattr(parent_module, child_name, new_module)
-        new_module.weight = old_module.weight
-        if old_module.bias is not None:
-            new_module.bias = old_module.bias
-        if getattr(old_module, "state", None) is not None:
-            new_module.state = old_module.state
-            new_module.to(old_module.weight.device)
+    def _replace_module(parent, target_name, new_module, target):
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/peft/tuners/ia3.py]: Done in [8b88401](https://github.com/huggingface/peft/pull/749/commits/8b88401eeafde6bceb4763ecf1530bf7acda80db)

File: src/peft/utils/save_and_load.py [hunk 28]
```diff
@@ -118,14 +127,74 @@ def set_peft_model_state_dict(model, peft_model_state_dict, adapter_name="defaul
...
+            Additional arguments to pass to the `hf_hub_download` method when loading from the HuggingFace Hub.
+    """
+    path = (
+        os.path.join(model_id, hf_hub_download_kwargs["subfolder"])
+        if hf_hub_download_kwargs.get("subfolder", None) is not None
+        else model_id
+    )
+
+    if device is None:
+        device = "cuda" if torch.cuda.is_available() else "cpu"
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/peft/utils/save_and_load.py]: Done in [84a56e0](https://github.com/huggingface/peft/pull/749/commits/84a56e096d0d7a743770cf5c9803556ab3f93f55)

File: src/peft/tuners/tuners_utils.py [hunk 29]
```diff
@@ -0,0 +1,240 @@
...
+
+        # For advanced developpers, if you want to attach multiple adapters to your
+        # model, just add a `peft_config` attribute to your model.
+        if not hasattr(self, "peft_config"):
+            self.peft_config = peft_config
+        else:
+            logger.info(
+                "Already found a `peft_config` attribute in the model. This will lead to having multiple adapters"
+                " in the model. Make sure to know what you are doing!"
+            )
```
Comment (pacman100) [inline src/peft/tuners/tuners_utils.py:74]: post this you have to add the new config in the `peft_config` dict, else `inject_adapter` will lead to key not found error

**Review by pacman100 — COMMENTED** [review]
Thank you for all the changes and bearing with me 🤗. Left a comment

File: src/peft/tuners/tuners_utils.py (hunk 29 above)
Comment (younesbelkada) [reply src/peft/tuners/tuners_utils.py:76]: Addressed it in  8053c88

File: src/peft/mapping.py [hunk 30]
```diff
@@ -89,10 +99,37 @@ def get_peft_model(model: PreTrainedModel, peft_config: PeftConfig, adapter_name
...
+        raise ValueError("`create_and_replace` does not support prompt learning and adaption prompt yet.")
+
+    if peft_config.peft_type not in PEFT_TYPE_TO_TUNER_MAPPING.keys():
+        raise ValueError(
+            f"`inject_adapter_in_model` does not support {peft_config.peft_type} yet. Please use `get_peft_model`."
+        )
+
+    tuner_cls = PEFT_TYPE_TO_TUNER_MAPPING[peft_config.peft_type]
+
+    peft_model = tuner_cls(model, peft_config, adapter_name=adapter_name)
```
Comment (patrickvonplaten) [inline src/peft/mapping.py:133]: If I went down the rabbit hole correctly [here](https://github.com/huggingface/peft/blob/ec267c644a9a9f05a7340a7cb23ed5a6a6090dd0/src/peft/tuners/lora.py#L202) it looks like this line will set `required_grad=False` for all weights of the model. For https://github.com/huggingface/transformers/pull/25077 this would mean that if a load lora weights of a model all weights of the model (except the lora weights) are frozen - I would not expect such a behavior here.

Also it seems like the peft config decides further down the road here: https://github.com/huggingface/peft/blob/ec267c644a9a9f05a7340a7cb23ed5a6a6090dd0/src/peft/tuners/lora.py#L203 if we're in inference mode or not - for the transformers implementation I would make try to not make any assumptions here depending on the config but leave it fully up to the user - so just not do anything (don't freeze any grads). 

Can we add a flag here that would allow for this ?

Also I'd maybe add a comment to make it a bit easier for the user to understand what this line does.
```suggestion
    # By instantiating a peft model we are injecting randomly initialized LoRA layers into the model's modules.
    peft_model = tuner_cls(model, peft_config, adapter_name=adapter_name)
```

Comment (younesbelkada) [reply src/peft/mapping.py:133]: Freezing the base model should be optional when you load a model I think you are right

However when you attach fresh new adapters usually (for 99% of the usecases) it is to train them, so maybe we should make a distinction

1- load a pretrained adapter --> not necessearly freeze the base model
2- add a new random adapter --> I would expect users to train the adapters --> freeze the base model

WDYT? @pacman100 @BenjaminBossan

Comment (patrickvonplaten) [reply src/peft/mapping.py:134]: That makes sense, I would maybe give the user full flexibility here then and add a new function argument:
`freeze_weights=True/False/"non-lora-only"` or something similar? I think when calling this from `transformers` we don't want all the model to be frozen, but for other use cases it could very much be the case I'd say

Comment (BenjaminBossan) [reply src/peft/mapping.py:134]: > 1- load a pretrained adapter --> not necessearly freeze the base model
> 2- add a new random adapter --> I would expect users to train the adapters --> freeze the base model

> That makes sense, I would maybe give the user full flexibility here then and add a new function argument

Personally, I would prefer to give the user an explicit argument, rather than (or in addition to) trying to guess based on the circumstances, so I would go with that option.

@patrickvonplaten: You bring up some good points. I will still merge the PR as is because Younes will be OoO for some time and we don't want this big refactor to become stale. We should be able to address your concerns in a follow up PR.

Comment (patrickvonplaten) [reply src/peft/mapping.py:134]: Sounds good! Agree that giving the user an explicit argument is the right option here
